packages/devtools_app/lib/src/timeline/timeline_flame_chart.dart

```diff
@@ -333,6 +413,119 @@ class FullTimelineFlameChartCanvas extends FlameChartCanvas<FullTimelineData> {
         0;
     return absoluteY - topOffset - (section * sectionSpacing);
   }
+
+  @override
+  void paintCallback(CanvasRenderingContext2D canvas, Rect visible) {
+    paintSections(canvas, visible);
+    paintRows(canvas, visible);
+    _paintAsyncGuidelines(canvas, visible);
+    paintTimelineGrid(canvas, visible);
+  }
+
+  void _paintAsyncGuidelines(CanvasRenderingContext2D canvas, Rect visible) {
+    final firstVerticalGuideline = _binarySearchForVerticalLineSegment(visible);
+    final firstHorizontalGuideline =
+        _binarySearchForHorizontalLineSegment(visible);
+
+    // Only modify the canvas style if we have any guidelines to paint.
+    if (firstHorizontalGuideline != null || firstVerticalGuideline != null) {
+      canvas
+        ..strokeStyle = colorToCss(guidelineColor)
+        ..lineWidth = guidelineWidth;
+    }
+
+    if (firstVerticalGuideline != null) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        verticalGuidelines,
+        firstVerticalGuideline,
+      );
+    }
+
+    if (firstHorizontalGuideline != null) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        horizontalGuidelines,
+        firstHorizontalGuideline,
+      );
+    }
+  }
+
+  void _paintGuidelines(
+    CanvasRenderingContext2D canvas,
+    Rect visible,
+    List<LineSegment> guidelines,
+    LineSegment first,
+  ) {
+    final guidelinesToPaint = guidelines
+        .sublist(guidelines.indexOf(first))
+        .where((line) => line.intersects(visible));
+    for (var line in guidelinesToPaint) {
+      canvas
+        ..beginPath()
+        ..moveTo(line.start.x, line.start.y + topOffset)
+        ..lineTo(line.end.x, line.end.y + topOffset)
+        ..closePath()
+        ..stroke();
+    }
+  }
+
+  LineSegment _binarySearchForVerticalLineSegment(Rect rect) {
+    final LineSegmentSearchCondition shouldSearchFirstHalf =
```

Comment:
why do we need these as checks? that seems like a bit of a code smell.
I would recommend adding a
LineSegmentData or LineSegmentCollection class that takes a List<LineSegment> or List<T extends LineSegment>
and provides the methods to do fast lookup within the list of line segments, etc.

Reply:
removed this code in favor of using `lowerBound` from package:collection